JoinColony/colonyDapp
# Wire up Safe transaction action page

## Description

This PR wires in the ActionPage when initiating one or multiple Safe Transactions. 

**New stuff** ✨

* DetailsWidgetSafeTransaction component to display safe transactions on the action page

**Changes** 🏗

* Wire in the SafeTransaction with the ActionsPage components 
* Update colonyAction resolver to accomodate SafeTransactions
* Add messages for SafeTransactions
* Adjust `ActionsPageFeedItemWithIPFS` to accomodate the way we're passing the annotation message in without a corresponding hash
* Move type defs to graphql type defs file

Resolves #3778, #4038

## To test
To test, set up your bridging environment (instructions copied below). Then initiate each safe tx individually, as well as safe transactions multiply. The subsequent actions pages should appear like in the designs. Test with and without an added annotation. Test with recipients selected from the dropdown and recipients entered manually. All should work properly. 

With respect to the contract interaction section, I have included an abi below that you can manually enter to test how it performs when given array and nested array function parameters. 

## Figma: What it's supposed to look like
**Note: approve call is not required**

CI: https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=137%3A9262
Raw tx: https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=136%3A8814
NFT Transfer: https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=126%3A9155
Transfer Funds: https://www.figma.com/file/LlVmeMQxHVA04evfrgZyuN/Safe-Control?node-id=126%3A9670

## Screenshots: How it looks

Transfer NFT:

**Note: if your NFT is missing their avatar, this [issue](https://github.com/JoinColony/colonyDapp/issues/3973) has been created to implement a fallback in the event the css cannot fetch the url, as has been happening to me recently since Rinkeby was deprecated. This will also likely be true for ERC20 tokens when transferring funds.**

![tnft](https://user-images.githubusercontent.com/64402732/196974922-6ef3ca14-ae8d-47a6-b4d4-ba2ffa5fed5a.png)

Transfer Funds:
![tf](https://user-images.githubusercontent.com/64402732/196974928-fead8dc3-dd97-4f4b-ade6-f3e5a914613c.png)

Raw Transaction:
![rawtx](https://user-images.githubusercontent.com/64402732/197176301-b8fe4cbf-3b94-44c5-8e46-f0fd44321781.png)

Contract Interaction:
![ci](https://user-images.githubusercontent.com/64402732/196974941-0c9e3a7f-5858-4598-a18f-cf4be5927974.png)

Multiple Tx:
![multi](https://user-images.githubusercontent.com/64402732/196974938-68844e56-4405-44df-b4e7-4a824909963e.png)

## Notes

@ArmandoGraterol 
1. Contract interaction, raw tx and multiple tx messages weren't specc'd. Please review & advise.
1. Contract and function contract refer to the same contract, just one shows the name and one the address -- can these fields be merged? To me, it reads as if they're referring to different contracts.
1. Also see how I'm handling nested arrays and if desired, I can open a separate issue to format the nested arguments (e.g. if a nested array contains uints, wrap those uints in the `Numeral` component)

![handlingArrays](https://user-images.githubusercontent.com/64402732/196932679-a238ebf2-475d-4d49-8cd4-b86b8cd94b20.png)

## Mock abi to test Contract Interaction section
```
[
	{
		"inputs": [
			{
				"internalType": "bytes4[]",
				"name": "a",
				"type": "bytes4[]"
			},
			{
				"internalType": "int8[][]",
				"name": "b",
				"type": "int8[][]"
			},
			{
				"internalType": "address[]",
				"name": "c",
				"type": "address[]"
			},
			{
				"internalType": "address[][]",
				"name": "d",
				"type": "address[][]"
			}
		],
		"name": "one",
		"outputs": [],
		"stateMutability": "nonpayable",
		"type": "function"
	}
]
```

## Bridging env instructions

0. have your normal dev env running. 
1. `cd src/lib/colonyNetwork` 
2. `git fetch` 
3. `git checkout maint/cross-chain-safe` 
4. `git submodule update --init` 
5. `yarn` 
6. `yarn provision:safe:contracts` 
7. `cd lib/safe-contracts` then `yarn` (ignore errors)
8. `cd ../..` (back into colonyNetwork) 
9. `bash scripts/start-bridging-environment.sh` 
10. `node scripts/start-bridging-environment.js` 
11. Copy the data in console into your .env and restart webpack (home bridge address, local safe address, local token address etc.)

```
# For testing Safe Control locally. Ensure you have a foreign chain running locally.

LOCAL_HOME_BRIDGE_ADDRESS=

LOCAL_FOREIGN_BRIDGE_ADDRESS=

LOCAL_SAFE_ADDRESS=

ZODIAC_BRIDGE_MODULE_ADDRESS=

LOCAL_ERC721_ADDRESS=

LOCAL_SAFE_TOKEN_ADDRESS=
```
